clang/docs/StandardCPlusPlusModules.rst

```diff
@@ -652,6 +652,141 @@ in the future. The expected roadmap for Reduced BMIs as of Clang 19.x is:
    comes, the term BMI will refer to the Reduced BMI and the Full BMI will only
    be meaningful to build systems which elect to support two-phase compilation.
 
+Experimental No Transitive Change
+---------------------------------
+
+This section is primarily for build system vendors. For end compiler users,
```

Comment:
This may have been discussed somewhere that I missed, please point me to that if that's the case.

I was putting a little more thought into it and am wondering what would be a way for the build systems to take advantage of that?

When compiling something that dependends on modules, the build system will need to check if any of the (transitive) dependencies of the modules changes, right? Because any BMI that was changed in the transitive set can potentially affect the result of the current compilations.

So even if direct dependencies did not change (because of this optimization), despite a change in the deeper transitive dependency, the build system still can't reuse the result and has to rerun the compile.

We only know that the output BMI wasn't affected by any of the transitive changes **after** we finish compiling it and can compare the outputs, right?
But the build system needs to know about it **before** the compilation happens to avoid recompilations.

I suspect I'm missing something, but don't know what...

Reply:
> This may have been discussed somewhere that I missed, please point me to that if that's the case.
> 
> I was putting a little more thought into it and am wondering what would be a way for the build systems to take advantage of that?
> 
> When compiling something that dependends on modules, the build system will need to check if any of the (transitive) dependencies of the modules changes, right? Because any BMI that was changed in the transitive set can potentially affect the result of the current compilations.
> 
> So even if direct dependencies did not change (because of this optimization), despite a change in the deeper transitive dependency, the build system still can't reuse the result and has to rerun the compile.

Yes, and this is the reason why it is experimental.

> 
> We only know that the output BMI wasn't affected by any of the transitive changes **after** we finish compiling it and can compare the outputs, right? But the build system needs to know about it **before** the compilation happens to avoid recompilations.
> 
> I suspect I'm missing something, but don't know what...

It is not about to avoid the re-compilation for that unchanged BMI, but for other TUs that only dependent on the unchanged BMI. For example,

```
// a.cppm
export module a;
// intentional empty
```

```
// b.cppm
export module b;
import a;
// intentional empty
```

```
// c.cpp
import b;
...
```

For this example, every time `a.cppm` changes, we need to recompile `b.cppm`. But if the BMI of module `B` doesn't change at all, it should be good to not recompile `c.cpp`. Do this make sense?

Reply:
That's what I thought is happening, thanks for confirming my intuition.
I think it'd be great to spell this out in the documentation, because I believe this mode of operation is something that the build system vendors might need to infer from the whole text right now.

Adding something like the following should do the trick:

```
We encourage build systems to add an experimental mode that
reuses the cached BMI when **direct** dependencies did not change,
even if **transitive** dependencies did change.
```

PS My 2 cents on Bazel: I suspect this actually goes against its design. Bazel has very strong capabilities to avoid recompilations when inputs don't change, but it relies on stable hashes for all inputs (including transitive dependencies). 
At the same time I suspect that other build systems (`CMake` + `ccache` / `sccache`) are more flexible in that regard.

PPS I am not a build system expert, so take that with a grain of salt. I'm also happy to loop in Bazel folks I work with to confirm or rebut my claims about Bazel's design, if that's useful.

Reply:
Yeah, added suggested text. 

For bazel, I think it might be fine since it should be configurable that whether not to recompile a target as far as I know.